src/compiler/checker.ts

```diff
@@ -15958,6 +15961,26 @@ namespace ts {
             return type;
         }
 
+        function instantiateReverseMappedType(type: ReverseMappedType, mapper: TypeMapper) {
+            const innerMappedType = instantiateType(type.mappedType, mapper);
+            if (!(getObjectFlags(innerMappedType) & ObjectFlags.Mapped)) {
+                return type;
+            }
+            const innerIndexType = instantiateType(type.constraintType, mapper);
+            if (!(innerIndexType.flags & TypeFlags.Index)) {
+                return type;
+            }
+            const instantiated = inferTypeForHomomorphicMappedType(
+                instantiateType(type.source, mapper),
+                innerMappedType as MappedType,
+                innerIndexType as IndexType
+            );

```

Comment:
So, the instantiation for a reverse mapped type inferred from some generic source is just another reverse mapped type for the instantiation of that source. Is that the right way to read this?

So, to walk through a slight simplification of Anders’ `unboxify` example:

```ts
type Boxified<T> = { [P in keyof T]: { value: T[P] } };

declare function unboxify<T>(obj: Boxified<T>): T;

function foo<U>(obj: { prop: { value: U }  }) {
    return unboxify(obj);
}

foo({ prop: { value: 10 } });
```

At the `unboxify(obj)` call, we infer a reverse mapped type from source `{ prop: { value: U } }` to target `T`. Later, at the `foo({ prop: { value: U } })` when we have a mapper from `U` to `number`, we instantiate the aforementioned reverse mapped type by making another reverse mapped type as if we were performing inference from source `{ prop: { value: number } }` to target `T`, which, by the actual reverse mapping mechanism (unchanged in this PR, and I don’t know where it lives or when it’s triggered) to resolve that reverse mapped type into `{ prop: number }`. Prior to this PR, basically the same process was happening, but we were trying to resolve the original reverse mapped type with the uninstantiated source, so we would have just gotten `{ prop: U }`.

I still don’t really understand the instantiations of `type.mappedType` and `type.constraintType`—in the examples I’ve walked through, those `instantiateType` calls end up just returning the original type passed to them. E.g., in the `unboxify` example, the mapped type and constraint type only have references to `T`, and the mapper only maps `U` (which is only relevant to the source), and it’s hard to imagine how the inner mapped type could contain any type variables you’d ever find in the source. Is this _only_ relevant to the special mappers you mentioned to Anders earlier?

Reply:
Or, let me ask a more specific question about `permissiveMapper` and the error scenarios you mentioned—what would be the consequences of, rather than early returning the input `type`, just proceeding to return the reverse mapped type with an instantiated source, but with `mappedType` and `constraintType` simply copied:

```ts
function instantiateReverseMappedType(type: ReverseMappedType, mapper: TypeMapper) {
     return inferTypeForHomomorphicMappedType(
         instantiateType(type.source, mapper),
         type.mappedType,
         type.constraintType
     ) || type;
}
```

I think this would have been my instinct, and I don’t understand what would go wrong, in what cases.

Reply:
Yeah, function type mappers that eliminate _all_ type variables are _pretty much_ the only ones that would affect those inner instantiations. The effect would be those type parameters would go unmapped, so we'd get inaccurate results when we use them. The means that, eg, we'd measure conditional type `extends` checks incorrectly when they involved type parameters at these positions within a reverse mapped type.